### ISSUE TITLE
Cannot construct a Request with a Request object that has already been used

### ISSUE THREAD

**Issue by vladimirmartsul** [pwa-builder/serviceworkers#16]
Hello

I'm trying to use serviceWorker2 (Offline copy of pages) and get error during pages navigation after successlully installing service worker:

Google Chrome 62.0.3202.89 (Official Build) (64-bit)
`Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'ServiceWorkerGlobalScope': Cannot construct a Request with a Request object that has already been used.
    at pwabuilder-sw.js:19
    at <anonymous>
(anonymous) @ pwabuilder-sw.js:19
Promise rejected (async)
(anonymous) @ pwabuilder-sw.js:26
17:00:08.506 `

Firefox 57.0 (64-bit)
`Service worker event waitUntil() was passed a promise that rejected with 'TypeError: Body has already been consumed.'.`

**Comment by go4site** [comment]
Same here. Have you found any solution?

**Comment by marie-dk** [comment]
I am getting the same error in Chrome 62.0.3202.94 (Official Build) (64-bit):
```
Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'ServiceWorkerGlobalScope': Cannot construct a Request with a Request object that has already been used.
    at pwabuilder-sw.js:38
```

Apparently fetch does not accept the same request twice... but I have no idea how to fix it.
https://stackoverflow.com/questions/35394924/why-does-this-code-fail-to-execute-fetch

**Comment by CymruUSA** [comment]
I have the same issue. Has anyone found a solution yet?

**Comment by boyofgreen** [comment]
Hello,  sorry no one was monitoring these issues until now, but we have already found the scenario where this issues arises, and we're testing a fix now.  I'll let you know when you can re-try

**Comment by CymruUSA** [comment]
Any update on this? Thank you.

**Comment by CymruUSA** [comment]
Or can you give us an idea of what scenario causes this.

**Comment by boyofgreen** [comment]
hi CymruUSA we are planning on pushing a fix out here for the next release.  Once it's in the repo, I'll let you know so you can try it for us too!

**Comment by CymruUSA** [comment]
Thank you @boyofgreen

**Comment by hamzahamidi** [comment]
I'm facing the same issue.

**Comment by CymruUSA** [comment]
@hamzahamidi We're just waiting for the fix to go out in the new release. @boyofgreen said it will be soon.

**Comment by CymruUSA** [comment]
@boyofgreen I know you said you had a dix but any idea when you'll be pushing the next release so we can get a fix on this? Cheers.

**Comment by boyofgreen** [comment]
Yes sorry for the delay, we will push this out before the end of the month.

**Comment by CymruUSA** [comment]
@boyofgreen Great news. I guess you've been busy at build.

**Comment by icarito** [comment]
I was getting `Cannot construct a Request with a Request object that has already been used.` and apparently was able to fix it with https://github.com/icarito/serviceworkers/commit/f17f892ba9f2be057aeffa133f01f243604ddc0c

**Comment by CymruUSA** [comment]
@boyofgreen It looks like the original issue has been fixed but now I get a Uncaught "(in promise) TypeError: Request failed" at sw.js:1. Should I open up a new issue? FYI, my sw.js looks like this;

//This is the service worker with the combined offline experience (Offline page + Offline copy of pages)

//Install stage sets up the offline page in the cache and opens a new cache
self.addEventListener('install', function(event) {
  event.waitUntil(preLoad());
});

var preLoad = function(){
  console.log('[PWA Builder] Install Event processing');
  return caches.open('pwabuilder-offline').then(function(cache) {
    console.log('[PWA Builder] Cached index and offline page during Install');
    return cache.addAll(['/offline.html', '/index.html']);
  });
}

self.addEventListener('fetch', function(event) {
  console.log('The service worker is serving the asset.');
  event.respondWith(checkResponse(event.request).catch(function() {
    return returnFromCache(event.request)}
  ));
  event.waitUntil(addToCache(event.request));
});

var checkResponse = function(request){
  return new Promise(function(fulfill, reject) {
    fetch(request).then(function(response){
      if(response.status !== 404) {
        fulfill(response)
      } else {
        reject()
      }
    }, reject)
  });
};

var addToCache = function(request){
  return caches.open('pwabuilder-offline').then(function (cache) {
    return fetch(request).then(function (response) {
      console.log('[PWA Builder] add page to offline'+response.url)
      return cache.put(request, response);
    });
  });
};

var returnFromCache = function(request){
  return caches.open('pwabuilder-offline').then(function (cache) {
    return cache.match(request).then(function (matching) {
     if(!matching || matching.status == 404) {
       return cache.match('offline.html')
     } else {
       return matching
     }
    });
  });
};

**Comment by CymruUSA** [comment]
Spoke too soon. It appears to be back. 
![uncaught](https://user-images.githubusercontent.com/35611890/40071464-d8b45b3c-583e-11e8-8fcb-74798ef47da6.jpg)
Uncaught (in promise) TypeError: Failed to execute 'clone' on 'Request': Request body is already used
    at sw.js:38
(anonymous) @ sw.js:38

**Comment by kdavis45** [comment]
Same issue as well. Any fix?

Uncaught (in promise) TypeError: Failed to execute 'fetch' on 'ServiceWorkerGlobalScope': Cannot construct a Request with a Request object that has already been used.

**Comment by englishextra** [comment]
That's because Chrome doesn't want to fetch files from root app folder. This behavior is an old one.

**Comment by webraptor** [comment]
If you want to use the same _Request_ object twice, you cannot. However, you can clone it using _.clone()_ method and use the original request and the cloned request in your code.

One use case would be when you want to return a request from the cache, or if it is missing from the cache from network, but at the same time update the cache, but not wait for the request to finish before serving the response from cache.

```
const clonedRequest = request.clone();

// trigger network update
fetch(request).then(function(response) { // will work
  caches.open(cacheConfig.name).then(function(cache) {
    cache.put(request, response);
  });
});

// respond from cache or fallback 2 network
e.respondWith(
  caches.match(request).then(function(response) {
    return response || fetch(request); // **will not work** since request has already been used
  })
);

// respond from cache or fallback 2 network
e.respondWith(
  caches.match(clonedRequest).then(function(response) {
    return response || fetch(clonedRequest); // **will work** since request has already been used
  })
);
```
